### ISSUE TITLE
More flexibility for the "Improve this doc" link

### ISSUE THREAD

**Issue by jmini** [docToolchain/docToolchain#856]
**Is your feature request related to a problem? Please describe.**

The "Improve this doc" link is great, but quite limited in the context of the usage with bitbucket.

<img width="317" alt="Screenshot with _Improve this doc_ link" src="https://user-images.githubusercontent.com/1222165/171624835-fe1b11fe-2459-4a1b-b307-84103832f2c5.png">

The problem is that the branch is added as query parameter at the end of the url with the `at` query parameter.

```
<bitbucket server>/projects/<project name>/repos/<repo name>/browse/<file-path>?at=<branch>
```

Example file `src/README.adoc` in the `TODO` project in the `something` repo in branch `my/branch`:
```
https://company.com/bitbucket/projects/TODO/repos/something/browse/src/README.adoc?at=refs%2Fmy%2Fbranch
```

Other example public bitbucket server (URL seems to be a little bit different):
https://bitbucket.org/atlassian/jira-rest-java-client/src/master/README?at=RAID-1043%2Ffix-automerge

**Describe the solution you'd like**
Ideal (not sure if this is possible) in the `docToolchainConfig.groovy` I would like to have a function that will be called.

Input: `String sourceFileName`
Output: `String` the url (that will become `href` value of the link)

Default implementation is: `(String sourceFileName) -> gitRepoUrl + "/" + sourceFileName`

That corresponds to:
https://github.com/docToolchain/docToolchain/blob/c0e3ceae51c786462d0340b8fb71eb97e39f2f3e/src/site/templates/rightcolumn.gsp#L7-L8

In this context, what would also be possible is to have different values depending if you build locally or a website.
Locally I like to do is to have a link to open in vscode:

```
vscode://file{file-absolute-with-leading-slash}
```
See: https://github.com/Microsoft/vscode-docs/blob/master/docs/editor/command-line.md#opening-vs-code-with-urls

But in this case `sourceFileName` should not be relative to the git project but absolute on the disk.

**Describe alternatives you've considered**

In `src/site/templates/rightcolumn.gsp` I am doing something like this:

```html
        <a href="${config.site_gitRepoUrl}/${sourceFileName}?at=refs%2Fheads%2F${java.net.URLEncoder.encode(config.site_branch, "UTF-8")}"
           target="_blank"><i class="fa fa-edit fa-fw"></i> Improve this doc</a>
```

But this means that I have to customize my `rightcolumn.gsp` in each of my projects.

**Comment by rdmueller** [comment]
hm.
So, we would have to replace the `rightcolumn.gsp` with something like this:

```
<a href="${config.site_gitRepoEditUrl(sourceFileName)}" target="_blank">
```

and define in the config something like this:

```
gitRepoEditUrl = { sourceFileName ->
    return gitRepoUrl + "/" + sourceFileName
}
```

and this could already do the trick. The config is a groovy file which allows us to execute arbitrary code and even define a function as a closure.

btw: do you know that you can zip your theme and reference it in your `dtcw` in order to have a central maintained company theme?

**Comment by jmini** [comment]
Is the addition of a function in `docToolchainConfig.groovy` considered as a breaking change?

Because if `rightcolumn.gsp` is updated (because they consume the released version) and users stick to an old `docToolchainConfig.groovy` version in their project, this will break something not?

**Comment by rdmueller** [comment]
yes, if we change it in the default theme, this will break things.

but we can check if the url is a String or a Closure:

short form:

```
<a href="${config.site_gitRepoEditUrl instanceof Closure?config.site_gitRepoEditUrl(sourceFileName):config.site_gitRepoEditUrl}" target="_blank">
```

long form:
```
<% if (config.site_gitRepoEditUrl instanceof Closure) { %>
  <a href="${config.site_gitRepoEditUrl(sourceFileName)}" target="_blank">
<% } else { %>
  <a href="${config.site_gitRepoEditUrl}" target="_blank">
<% } %>
```

I am quite happy that we have this flixibility through the gsp templates :-)

**Comment by jmini** [comment]
_I have the feeling I will learn a lot of Groovy while working on those topics with you_

**Comment by rdmueller** [comment]
that would be great! :-D

**Comment by jmini** [comment]
_Sorry for the delay_

Using a closure in the `docToolchainConfig.groovy` file does not seems to work:
```groovy
microsite.with {
    //..
    gitRepoEditUrl = { sourceFileName ->
        return "https://github.com/jmini/dtc-experiments/edit/${branch}/src/docs/${sourceFileName}"
    }
}
```
See https://github.com/jmini/dtc-experiments/commit/14e3df5be656926459f775730bf2d0f0c57ff40a

Is failing locally with a `cannot be serialized` error:
```
* What went wrong:
Execution failed for task ':bake'.
> Unable to store input properties for task ':bake'. Property 'configuration' with value '{asciidoctor.option.requires=asciidoctor-diagram, ... , site.branch=docToolchain-issue-856, site.gitRepoEditUrl=script16559734593161743290031$_run_closure1$_closure9@1e115c7, site.landingPage=landingpage.gsp, site.menu=, asciidoctor.attributes=[...]}' cannot be serialized.
```

On the topic I found this very old discussion:
https://stackoverflow.com/questions/9649073/serialization-of-closures-in-groovy

But I have no idea how to make this work
